### DISCUSSION TITLE
A little bit of care for making documentation for building from source.

### DISCUSSION THREAD

**Discussion by donaastor** [rakshasa/rtorrent#1323] [General]
I understand the developers struggle with finding time for developing, let alone writing documentation, but I hope that this very thing is important enough so I, with due respect, would kindly ask the developers to make a little update on it.

Current state is a bit broken: some links to "how to build/install" don't work and some are linking backwards in a loop. The only actual guide is inside the README file. It is not really explained and the scripts are long and it takes a lot of time to figure out what they actually do. I don't find it really friendly or easy for checking for malicious behavior (as in the spirit of open source, I am not hinting anything). As of libtorrent, there isn't even a source archive file, nothing except the github main branch and the tarball from the rtorrent page. There are no words on how to build and install libtorrent, maybe it should be implied that `./configure; make; make install` is to be executed.

Please make a short file explaining how to build libtorrent and rtorrent from source and a little bit of what each command does. I have already spent a few hours trying to figure everything out, but hadn't succeeded.

**Comment by kannibalox** [comment]
The wiki is freely editable, I've gone ahead and fleshed out this page a little more: https://github.com/rakshasa/rtorrent/wiki/Installing#installing, but feel free to clarify anything that is different and/or incorrect for your build system. It could certainly use more love, and the best time to get something documented is when someone is trying to come up to speed for the first time.

> I don't find it really friendly or easy for checking for malicious behavior (as in the spirit of open source, I am not hinting anything).

I don't want to discourage your efforts to stay secure, but whether something compiles easily is no indication of whether it's malicious. You'd need to understand and read the code yourself to really determine that. I certainly agree it could be more user-friendly

> Current state is a bit broken: some links to "how to build/install" don't work and some are linking backwards in a loop.

Do you have any specific examples for this?

**Reply by donaastor** [reply]
I see a solid update, thank you really for doing it! A quick question, `autoreconf -fi`, `./configure` and `make` only update files locally (in the same directory) and `make install` actually copies them to the root filesystem?

;

I agree with all your points, I will respond in this order:
> Do you have any specific examples for this?

Yes, I was referring to the community wiki, [here](https://rtorrent-docs.readthedocs.io/en/latest/installation.html#installing-from-source), the first link loops back to github, the next 2 links don't explain building, just installing, the 4th link doesn't work, 5th takes forever to respond and the last one responds with "page not found". I understand it is the *community* based wiki.
> It could certainly use more love

Your update is truly helpful, I will test it all out this weekend (I am really packed until then) and I was gonna offer to write out the steps with explanations myself, but I see you kinda already did that :D Regardless of that, I'll share my understandings in this thread, why not.
> I don't want to discourage your efforts to stay secure

Thank you; I can read C++ just fine, but in the top-down approach, where I'd start with compilation scripts, I get stuck as I have 0 clues what they do. All I know is `g++`/`clang` and their options.

**Regarding motivation**, even though I like keeping an eye on the security, it wasn't my main focus here. I wanted to understand compilation steps so that I'd know where do play with optimization flags and what files were created and how to install them manually. All in the goal of testing out the new `v0.10.0`. I know the last version had some bugs and Arch Linux repositories don't offer the new version.

**Reply by kannibalox** [reply]
> I see a solid update, thank you really for doing it! A quick question, `autoreconf -fi`, `./configure` and `make` only update files locally (in the same directory) and `make install` actually copies them to the root filesystem?

Exactly, if you want to change where `make install` places them, you can pass `--prefix=<path>` to configure, e.g. `./configure --prefix=$HOME/rtorrent/`, but you may need to pass additional flags to rtorrent for it to locate the non-standard libtorrent install path, e.g.
`PKG_CONFIG_PATH="$HOME/rtorrent/lib/pkgconfig"  ./configure --prefix=$HOME/rtorrent/`

> Yes, I was referring to the community wiki, [here](https://rtorrent-docs.readthedocs.io/en/latest/installation.html#installing-from-source), the first link loops back to github, the next 2 links don't explain building, just installing, the 4th link doesn't work, 5th takes forever to respond and the last one responds with "page not found". I understand it is the _community_ based wiki.

It's unfortunate but the person responsible for creating that organization/site hasn't responded to attempts to get it updated in a while. I've temporarily switched it out to my fork of the rtorrent-docs repo which at the very least removes the dead links.

**Reply by donaastor** [reply]
I can confirm that your updated steps on the installation page work as expected. I haven't tried building `rtorrent` with `libtorrent` in custom directory, I am still not sure how to call `make` for `rtorrent` in that scenario. Anyway, Arch Linux repo now packages the new version (0.10.0) so I simply installed that. This thread can probably be closed now. Thank you for your quick replies!
